shared-bindings/uvc/__init__.c

```diff
@@ -0,0 +1,116 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/obj.h"
+#include "py/runtime.h"
+
+#include "shared-bindings/uvc/__init__.h"
+#include "shared-bindings/uvc/UVCFramebuffer.h"
+#include "shared-module/uvc/__init__.h"
+
+//| """Allows streaming bitmaps to a host computer via USB UVC
+//|
+//| This interface is experimental and may change without notice even in stable
+//| versions of CircuitPython."""
+//|
+//| bitmap: displayio.Bitmap
+//| """The content of this bitmap image is streamed to the host computer. To change the
+//| displayed image, modify the bitmap contents.
+//|
+//| The bitmap content is in RGB565_SWAPPED format. It is converted to "YUYV" format before
+//| transmission to the host computer. The "YUYV" format stores a "Y" (luminance) sample
+//| for every pixel, but a chromonance (UV) value only for each pair of pixels; due to this,
+//| the color of adjacent pixels cannot be fully controlled. For instance, there is no way
+//| for the first pixel to be red and the second pixel to be green. However, they can both
+//| be different shades of red.
+//| """
+//|
+
+//| def enable_framebuffer(width: int, height: int) -> None:
+//|     """Enable the UVC gadget, setting the given width & height
+//|
+//|     During ``boot.py``, this function may be used to enable the UVC gadget.
+//|
+//|     Width is rounded up to a multiple of 2.
+//|
+//|     After boot.py completes, the framebuffer will be allocated. Total storage
+//|     of 4×``width``×``height`` bytes is required, reducing the amount available
+//|     for Python objects. If the allocation fails, a note will be written in boot.py
+//|     and UVC will be disabled"""
+//|
+
+STATIC mp_obj_t uvc_enable_framebuffer(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_width, ARG_height };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_width, MP_ARG_REQUIRED | MP_ARG_INT, { .u_int = 0 } },
+        { MP_QSTR_height, MP_ARG_REQUIRED | MP_ARG_INT, { .u_int = 0 } },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    // (but note that most devices will not be able to allocate this much memory.
+    uint32_t width = mp_arg_validate_int_range(args[ARG_width].u_int, 0, 32767, MP_QSTR_width);
+    uint32_t height = mp_arg_validate_int_range(args[ARG_height].u_int, 0, 32767, MP_QSTR_height);
+    if (!shared_module_uvc_enable(width, height)) {
+        mp_raise_RuntimeError(MP_ERROR_TEXT("Cannot change USB devices now"));
+    }
+
+    return mp_const_none;
+};
+STATIC MP_DEFINE_CONST_FUN_OBJ_KW(uvc_enable_framebuffer_obj, 0, uvc_enable_framebuffer);
+
+//|
+//| def disable() -> None:
```

Comment:
```suggestion
//| def disable_framebuffer() -> None:
```

I'd probably add `_framebuffer` so it's symmetric.

Reply:
I gave them asymmetric names because I hope we add support for MJPEG over UVC as well. There would be just one "disable", but there would be "enable_mjpeg". This mode would take a pixel size and possibly a maximum frame data size argument.

Reply:
Removing for now, since it's redundant.

Reply:
I'm ok leaving the plain `disable` too.